crates/sui-json-rpc/src/balance_changes.rs

```diff
@@ -182,55 +182,49 @@ impl<P> ObjectProviderCache<P> {
         }
     }
 
-    pub fn new_with_cache(
-        provider: P,
-        written_objects: BTreeMap<ObjectID, (ObjectRef, Object, WriteKind)>,
-    ) -> Self {
-        let mut object_cache = BTreeMap::new();
-        let mut last_version_cache = BTreeMap::new();
+    pub fn insert_objects_into_cache(&mut self, objects: Vec<Object>) {
+        let object_cache = self.object_cache.get_mut();
+        let last_version_cache = self.last_version_cache.get_mut();
 
-        for (object_id, (object_ref, object, _)) in written_objects {
-            let key = (object_id, object_ref.1);
+        for object in objects {
+            let object_id = object.id();
+            let version = object.version();
+
+            let key = (object_id, version);
             object_cache.insert(key, object.clone());
 
             match last_version_cache.get_mut(&key) {
                 Some(existing_seq_number) => {
-                    if object_ref.1 > *existing_seq_number {
-                        *existing_seq_number = object_ref.1
+                    if version > *existing_seq_number {
+                        *existing_seq_number = version
                     }
                 }
                 None => {
-                    last_version_cache.insert(key, object_ref.1);
+                    last_version_cache.insert(key, version);
                 }
             }
         }
-
-        Self {
-            object_cache: RwLock::new(object_cache),
-            last_version_cache: RwLock::new(last_version_cache),
-            provider,
-        }
     }
 
-    pub fn new_with_output_objects(provider: P, output_objects: Vec<Object>) -> Self {
+    pub fn new_with_cache(
```

Comment:
The name of this function is a bit confusing, also could it simply call `insert_objects_into_cache` over the objects from the map?

Reply:
This is just a bad looking diff, i didn't introduce that function (nor touch it) in this patch